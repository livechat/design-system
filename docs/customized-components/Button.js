import React from 'react';
import PropTypes from 'prop-types';
import { Button as DSButton } from '../../src/index';

function Button({ name, className, onClick, active, children }) {
  return (
    <DSButton
      className={className}
      onClick={onClick}
      name={name}
      primary={active}
      size="compact"
      style={{
        margin: '10px 0 0',
        marginBottom: active ? '5px' : '0'
      }}
    >
      {children}
    </DSButton>
  );
}

Button.propTypes = {
  name: PropTypes.string,
  className: PropTypes.string,
  onClick: PropTypes.func,
  active: PropTypes.bool,
  children: PropTypes.node
};

export default Button;
