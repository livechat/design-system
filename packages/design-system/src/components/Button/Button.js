import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './style.css';

const cx = classNames.bind(styles);
const acceptedSizes = ['large', 'compact'];

const Button = props => {
  const {
    children,
    primary,
    destructive,
    disabled,
    loading,
    size,
    fullWidth,
    submit,
    accessibilityLabel,
    ariaControls,
    ariaExpanded,
    ...buttonProps
  } = props;

  const isDisabled = disabled || loading;
  const type = submit ? 'submit' : 'button';
  let buttonType = null;

  if (primary) {
    buttonType = 'primary';
  } else if (destructive) {
    buttonType = 'destructive';
  }

  return (
    <button
      {...buttonProps}
      type={type}
      disabled={isDisabled}
      role={loading ? 'alert' : undefined}
      aria-busy={loading ? true : undefined}
      aria-label={accessibilityLabel}
      aria-controls={ariaControls}
      aria-expanded={ariaExpanded}
      className={cx({
        btn: true,
        'btn--full-width': fullWidth,
        'btn--primary': buttonType === 'primary',
        'btn--destructive': buttonType === 'destructive',
        [`btn--${size}`]: acceptedSizes.some(s => s === size)
      })}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.node.isRequired,
  id: PropTypes.string,
  /**
   * Type of button
   */
  primary: PropTypes.bool,
  /**
   * Type of button
   */
  destructive: PropTypes.bool,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  /**
   * Size of button
   */
  size: PropTypes.oneOf(['compact', 'large']),
  /**
   * Sets button width to max-width=320px
   */
  fullWidth: PropTypes.bool,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  submit: PropTypes.bool,
  className: PropTypes.string,
  accessibilityLabel: PropTypes.string,
  ariaControls: PropTypes.string,
  ariaExpanded: PropTypes.string
};

export default Button;
