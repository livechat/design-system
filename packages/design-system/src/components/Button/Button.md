### Primary Button

Use when there is an action that is more important than any other actions and you need to draw attention to it. Don’t use more than one primary button per screen to avoid confusing users. Not every screen needs a primary button.
```js
<Button primary>Primary Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button primary>Primary Button</Button>
  </ComponentHtmlMarkup>
```
### Basic Button

The standard button for most use cases. Only use another style if a button requires more or less visual attention.
```js
  <Button>Basic Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button>Basic Button</Button>
  </ComponentHtmlMarkup>
```
### Destructive Button

Use as a final confirmation for a destructive action such as deleting. These are found mostly in confirmation modals.
```js
  <Button destructive>Destructive Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button destructive>Destructive Button</Button>
  </ComponentHtmlMarkup>
```
### Button Sizes
#### Compact
Use in a table or list where space is tight.
```js
  <Button size="compact" primary>Compact Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button size="compact" primary>Compact Button</Button>
  </ComponentHtmlMarkup>
```
#### Regular
```js
  <Button primary>Regular Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button primary>Regular Button</Button>
  </ComponentHtmlMarkup>
```
#### Large
Use for the main call to action in empty states or for calls to action shown with large illustrations.
```js
<Button size="large" primary>Large Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button size="large" primary>Large Button</Button>
  </ComponentHtmlMarkup>
```
#### Full-width

Button expanded to the full width to fill the parent container (max-width: 320px)
```js
  <Button fullWidth primary>Full-width Button</Button>
```
```js noeditor
  <ComponentHtmlMarkup>
    <Button fullWidth primary>Full-width Button</Button>
  </ComponentHtmlMarkup>
```

### Best practices
 - Prioritize most important action; don’t use more than 1 primary button per screen.
 - Position buttons in consistent locations in the UI.
 - Use short yet descriptive labels ({verb} + {noun}).
 - Use strong verbs to encourage action.
 - Avoid unnecessary words and articles such as the, an, or a.