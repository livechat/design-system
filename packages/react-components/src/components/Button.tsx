import * as React from 'react';
import cx from 'classnames';
import { Loader } from './Loader';

export type ButtonSize = 'compact' | 'medium' | 'large';

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  kind?: 'basic' | 'primary' | 'secondary' | 'destructive' | 'text';
  size?: ButtonSize;
  disabled?: boolean;
  loading?: boolean;
  fullWidth?: boolean;
  loaderLabel?: string;
  icon?: React.ReactElement;
  iconPosition?: 'left' | 'right';
}

const baseClass = 'lc-btn';

export const Button: React.FC<ButtonProps> = ({
  loading = false,
  disabled = false,
  type = 'button',
  fullWidth = false,
  kind = 'basic',
  size = 'medium',
  icon = null,
  iconPosition = 'left',
  loaderLabel,
  className,
  children,
  ...props
}) => {
  const isDisabled = loading || disabled;

  const mergedClassNames = cx(
    className,
    baseClass,
    `${baseClass}--${kind}`,
    `${baseClass}--${size}`,
    {
      [`${baseClass}--loading`]: loading,
      [`${baseClass}--full-width`]: fullWidth,
      [`${baseClass}--icon-only`]: !children && icon,
    }
  );

  return (
    <button
      className={mergedClassNames}
      disabled={isDisabled}
      type={type}
      {...props}
    >
      {loading && (
        <Loader
          size="small"
          label={loaderLabel}
          className={`${baseClass}__loader`}
        />
      )}
      {icon &&
        React.cloneElement(icon, {
          className: cx(
            `${baseClass}__icon`,
            `${baseClass}__icon--${iconPosition}`
          ),
          disabled,
        })}
      <div className={`${baseClass}__content`}>{children}</div>
    </button>
  );
};
