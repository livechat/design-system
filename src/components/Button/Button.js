import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './style.scss';
import getMergedClassNames from '../../utils/getMergedClassNames';
import Loader from '../Loader';

const cx = classNames.bind(styles);
const acceptedSizes = ['large', 'compact'];

const Button = React.forwardRef((props, ref) => {
  const {
    children,
    primary,
    destructive,
    disabled,
    loading,
    size,
    fullWidth,
    submit,
    accessibilityLabel,
    ariaControls,
    ariaExpanded,
    className,
    type: htmlType,
    ...buttonProps
  } = props;

  const isDisabled = disabled || loading;
  const type = submit ? 'submit' : htmlType || 'button';
  let buttonType = null;

  if (primary) {
    buttonType = 'primary';
  } else if (destructive) {
    buttonType = 'destructive';
  }

  const baseClass = 'btn';
  const mergedClassNames = getMergedClassNames(
    cx({
      [baseClass]: true,
      [`${baseClass}--disabled`]: disabled,
      [`${baseClass}--loading`]: loading,
      [`${baseClass}--full-width`]: fullWidth,
      [`${baseClass}--primary`]: buttonType === 'primary',
      [`${baseClass}--destructive`]: buttonType === 'destructive',
      [`${baseClass}--${size}`]: acceptedSizes.some(s => s === size)
    }),
    className
  );

  return (
    <button
      {...buttonProps}
      ref={ref}
      className={mergedClassNames}
      type={type}
      disabled={isDisabled}
      role={loading ? 'alert' : undefined}
      aria-busy={loading ? true : undefined}
      aria-label={accessibilityLabel}
      aria-controls={ariaControls}
      aria-expanded={ariaExpanded}
    >
      {loading && (
        <div className={styles[`${baseClass}__loader`]}>
          <Loader />
        </div>
      )}
      {children}
    </button>
  );
});

Button.propTypes = {
  children: PropTypes.node.isRequired,
  id: PropTypes.string,
  /**
   * Type of button
   */
  primary: PropTypes.bool,
  /**
   * Type of button
   */
  destructive: PropTypes.bool,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  /**
   * Size of button
   */
  size: PropTypes.oneOf(['compact', 'large']),
  /**
   * Sets button width to max-width=320px
   */
  type: PropTypes.string,
  fullWidth: PropTypes.bool,
  onClick: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  submit: PropTypes.bool,
  className: PropTypes.string,
  accessibilityLabel: PropTypes.string,
  ariaControls: PropTypes.string,
  ariaExpanded: PropTypes.string
};

export default Button;
