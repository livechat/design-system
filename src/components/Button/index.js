export { default as Button } from './Button';
export { default as ButtonWithLoader } from './ButtonWithLoader';
