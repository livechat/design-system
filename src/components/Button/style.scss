@import '../../scss/variables/colors';

.btn {
  max-width: 320px;
  position: relative;
  align-items: center;
  justify-content: center;
  user-select: none;
  text-decoration: none;
  transition-property: opacity, border, box-shadow;
  transition-duration: 200ms;
  transition-timing-function: cubic-bezier(0.64, 0, 0.35, 1);
  border-width: 1px;
  border-style: solid;
  text-align: center;
  cursor: pointer;
  border-radius: 4px;
  font-weight: 600;
  display: inline-flex;
  background-color: #fff;
  color: $font-primary-color;
  border-color: $border-primary-color;
  font-size: 15px;
  padding: 0 16px;
  min-width: 36px;
  min-height: 36px;
  line-height: 36px;

  &:focus,
  &:hover {
    outline: 0;
    opacity: 0.8;
  }

  &:disabled {
    opacity: 0.65;
    cursor: not-allowed;
  }

  &--full-width {
    display: flex;
    width: 100%;
  }

  &--primary {
    background-color: $primary-color;
    color: #fff;
    border-color: $primary-color;

    &:focus {
      box-shadow: 0 0 0 1px $primary-color;
    }
  }

  &--destructive {
    background-color: $error-color;
    color: #fff;
    border-color: $error-color;

    &:focus {
      box-shadow: 0 0 0 1px $error-color;
    }
  }

  &--large {
    font-size: 15px;
    padding: 0 24px;
    min-width: 42px;
    min-height: 42px;
    line-height: 42px;
  }

  &--compact {
    font-size: 14px;
    padding: 0 16px;
    min-width: 32px;
    min-height: 32px;
    line-height: 32px;
  }

  .icon {
    color: $primary-color;
    margin-right: 7px;
  }
}
