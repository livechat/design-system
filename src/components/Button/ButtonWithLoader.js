import * as React from 'react';
import classNames from 'classnames/bind';
import { TransitionGroup, CSSTransition } from 'react-transition-group';
import styles from './style.scss';
import getMergedClassNames from '../../utils/getMergedClassNames';
import Loader from '../Loader';
import Button from './Button';

const cx = classNames.bind(styles);

const LOADER_ANIMATION_TIME = 200;

const ButtonWithLoader = props => {
  const { children, disabled, loading, className } = props;

  const baseClass = 'btn';
  const mergedClassNames = getMergedClassNames(
    cx({
      [`${baseClass}-with-loader`]: true,
      [`${baseClass}-with-loader--disabled`]: disabled,
      [`${baseClass}-with-loader--loading`]: loading
    }),
    className
  );

  return (
    <Button {...props} className={mergedClassNames}>
      <TransitionGroup component={null}>
        {loading && (
          <CSSTransition
            classNames={{
              enter: styles[`${baseClass}__loader-enter`],
              enterActive: styles[`${baseClass}__loader-enter-active`],
              exit: styles[`${baseClass}__loader-exit`],
              exitActive: styles[`${baseClass}__loader-exit-active`]
            }}
            timeout={LOADER_ANIMATION_TIME}
          >
            <div className={styles[`${baseClass}__loader`]}>
              <Loader />
            </div>
          </CSSTransition>
        )}
      </TransitionGroup>
      <span
        className={cx({
          [`${baseClass}-with-loader__content`]: true,
          [`${baseClass}-with-loader__content--hidden`]: loading
        })}
      >
        {children}
      </span>
    </Button>
  );
};

export default ButtonWithLoader;
