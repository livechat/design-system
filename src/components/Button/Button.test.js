import * as React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';
import ButtonWithLoader from './ButtonWithLoader';

const clickFn = jest.fn();
const focusFn = jest.fn();
const blurFn = jest.fn();

describe('Button', () => {
  it('should render primary button', () => {
    const component = shallow(
      <Button destructive primary>
        Click me
      </Button>
    );

    expect(component).toMatchSnapshot();
  });

  it('should render primary button with loader', () => {
    const component = shallow(
      <ButtonWithLoader primary loading>
        Click me
      </ButtonWithLoader>
    );

    expect(component).toMatchSnapshot();
  });

  it('on click, focus and blur should call passed in props functions', () => {
    const component = shallow(
      <Button onClick={clickFn} onFocus={focusFn} onBlur={blurFn}>
        Click me
      </Button>
    );
    const button = component.find('button');

    button.simulate('click');
    expect(clickFn).toHaveBeenCalled();

    button.simulate('focus');
    expect(focusFn).toHaveBeenCalled();

    button.simulate('blur');
    expect(blurFn).toHaveBeenCalled();
  });
});
