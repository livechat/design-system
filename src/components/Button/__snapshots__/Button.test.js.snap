// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render primary button 1`] = `
<button
  className="btn btn--primary"
  type="button"
>
  Click me
</button>
`;
